File: c/yacht/yacht.c
#include <stdbool.h>
#include <stdio.h>
#include "yacht.h"


int score(dice_t dice, category_t category)
{
    int counts[7] = {0};
    int total_sum = 0;
    int score = 0;

    bool has_three = false;
    bool has_two = false;

    for(int i = 0; i < 5; i++) {
        counts[dice.faces[i]]++;
        total_sum += dice.faces[i];
    }

    switch(category)
    {
        case ONES:
        case TWOS:
        case THREES:
        case FOURS:
        case FIVES:
        case SIXES:
            return counts[category] * category;

        case FULL_HOUSE:
            for(int i = ONES; i <= SIXES; i++) {
                if(counts[i] == 3) has_three = true;
                if(counts[i] == 2) has_two = true;
            }
            return (has_three && has_two) ? total_sum : 0;

        case FOUR_OF_A_KIND:
            for(int i = ONES; i <= SIXES; i++)
                if(counts[i] >= 4) return i * 4;
            break;

        case LITTLE_STRAIGHT:
            if(counts[1] == 1 && counts[2] == 1 && counts[3] == 1 && counts[4] == 1 && counts[5] == 1) return 30;
            break;

        case BIG_STRAIGHT:
            if(counts[2] == 1 && counts[3] == 1 && counts[4] == 1 && counts[5] == 1 && counts[6] == 1) return 30;
            break;
        
        case CHOICE:
            return total_sum;
            break;
        
        case YACHT:
            for(int i = ONES; i <= SIXES; i++)
                if(counts[i] == 5) return 50;
            break;
    }
    
    return score;
}

File: elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(:ones, dice), do: count_score(dice, 1)
  def score(:twos, dice), do: count_score(dice, 2)
  def score(:threes, dice), do: count_score(dice, 3)
  def score(:fours, dice), do: count_score(dice, 4)
  def score(:fives, dice), do: count_score(dice, 5)
  def score(:sixes, dice), do: count_score(dice, 6)

  def score(:choice, dice), do: Enum.sum(dice)

  def score(:yacht, dice) do
    if length(Enum.uniq(dice)) == 1, do: 50, else: 0
  end

  def score(:little_straight, dice) do
    if Enum.sort(dice) == [1, 2, 3, 4, 5], do: 30, else: 0
  end

  def score(:big_straight, dice) do
    if Enum.sort(dice) == [2, 3, 4, 5, 6], do: 30, else: 0
  end

  def score(:full_house, dice) do
    dice
    |> Enum.frequencies()
    |> Map.values()
    |> Enum.sort()
    |> case do
      [2, 3] -> Enum.sum(dice)
      _ -> 0
    end
  end

  def score(:four_of_a_kind, dice) do
    dice
    |> Enum.frequencies()
    |> Enum.find(fn {_face, count} -> count >= 4 end)
    |> case do
      {face, _count} -> face * 4
      nil -> 0
    end
  end

  defp count_score(dice, value) do
    value * Enum.count(dice, &(&1 == value))
  end
end
